pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,824 @@
+                qbx_forced_limit=orig_int_g.qbx_forced_limit,
+                target_kernel=orig_int_g.target_kernel.replace_base_kernel(
+                    int_g.target_kernel))
+        for int_g in int_gs}
+
+    substitutor = IntGSubstitutor(replacements)
+    return substitutor(expr)
+
+
+def merge_two_int_gs(int_g_1, int_g_2):
+    kernel_arguments = merge_kernel_arguments(int_g_1.kernel_arguments,
+            int_g_2.kernel_arguments)
+    source_kernels = int_g_1.source_kernels + int_g_2.source_kernels
+    densities = int_g_1.densities + int_g_2.densities
+
+    return int_g_1.copy(
+        source_kernels=tuple(source_kernels),
+        densities=tuple(densities),
+        kernel_arguments=kernel_arguments,
+    )
+
+
+class IntGCoefficientCollector(CoefficientCollector):
+    def __init__(self):
+        super().__init__({})
+
+    def map_int_g(self, expr):
+        return {expr: 1}
+
+    def map_algebraic_leaf(self, expr, *args, **kwargs):
+        return {1: expr}
+
+    handle_unsupported_expression = map_algebraic_leaf
+
+
+def evalf(expr, prec=100):
+    """evaluate an expression numerically using ``prec``
+    number of bits.
+    """
+    from sumpy.symbolic import USE_SYMENGINE
+    if USE_SYMENGINE:
+        return expr.n(prec=prec)
+    else:
+        import sympy
+        dps = int(sympy.log(2**prec, 10))
+        return expr.n(n=dps)
+
+
+@memoize_on_first_arg
+def _get_base_kernel_matrix(base_kernel, order=None):
+    dim = base_kernel.dim
+
+    pde = base_kernel.get_pde_as_diff_op()
+    if order is None:
+        order = pde.order
+
+    if order > pde.order:
+        raise NotImplementedError(f"order ({order}) cannot be greater than the order"
+                         f"of the PDE ({pde.order}) yet.")
+
+    mis = sorted(gnitstam(order, dim), key=sum)
+    # (-1, -1, -1) represent a constant
+    mis.append((-1, -1, -1))
+
+    if order == pde.order:
+        pde_mis = [ident.mi for eq in pde.eqs for ident in eq.keys()]
+        pde_mis = [mi for mi in pde_mis if sum(mi) == order]
+        logger.debug(f"Removing {pde_mis[-1]} to avoid linear dependent mis")
+        mis.remove(pde_mis[-1])
+
+    rand = np.random.randint(1, 100, (dim, len(mis)))
+    sym_vec = make_sym_vector("d", dim)
+
+    base_expr = base_kernel.get_expression(sym_vec)
+
+    mat = []
+    for rand_vec_idx in range(rand.shape[1]):
+        row = []
+        for mi in mis[:-1]:
+            expr = base_expr
+            for var_idx, nderivs in enumerate(mi):
+                if nderivs == 0:
+                    continue
+                expr = expr.diff(sym_vec[var_idx], nderivs)
+            replace_dict = dict(
+                (k, v) for k, v in zip(sym_vec, rand[:, rand_vec_idx])
+            )
+            eval_expr = evalf(expr.xreplace(replace_dict))
+            row.append(eval_expr)
+        row.append(1)
+        mat.append(row)
+
+    mat = sym.Matrix(mat)
+    L, U, perm = mat.LUdecomposition()
```

Comment:
I keep getting an intermittent `RuntimeError: Matrix is rank deficient` here when trying to compute the velocity gradient in 2D with the biharmonic-based Stokeslet.

Running this in a loop should hopefully fail
https://gist.github.com/alexfikl/81cd88cccd163d51939164c834046a68

Reply:
This is with `symengine`, with `sympy` it seems to fail below when doing the `chop`.

Reply:
I can't seem to reproduce this. Can you print the matrix and post it?

Reply:
Ah, just stared at the matrix for a second and the `-2` and `-4` rows are the same!